### PR TITLE
Reset POSHOLD target when GCS modifier is reset

### LINKED ISSUE #300: Switching from GCS_NAV to POSHOLD does not use current position
I have been playing around with the GCS_NAV mode to update the 'special waypoint' for a follow me functionality. I notice that when I switch off the GCS_NAV mode (returning to normal POSHOLD) the quad will keep moving to the 'special waypoint' location. I think it should instead set the current position as the desired position. One of the main reasons for having a distinct GCS_NAV mode was to avoid this behavior, because (to quote eziosoft) "The user may want to have PH ie in panic situation." #103

### PR DESCRIPTION
**Pull request by digitalentity**
Closes #300

### CODE REVIEW

**Comment by digitalentity** [comment]
@DzikuVx, @iforce2d can you review and if possible - test?

File: src/main/flight/navigation_rewrite.c [hunk 1]
```diff
@@ -731,7 +753,16 @@ static navigationFSMEvent_t navOnEnteringState_NAV_STATE_POSHOLD_3D_IN_PROGRESS(
 {
     UNUSED(previousState);
 
-     // If we enable terrain mode and surface offset is not set yet - do it
+    // If GCS was disabled - reset 2D pos setpoint
+    if (posControl.flags.isGCSAssistedNavigationReset) {
+        t_fp_vector targetHoldPos;
+        calculateInitialHoldPosition(&targetHoldPos);
+        setDesiredPosition(&posControl.actualState.pos, posControl.actualState.yaw, NAV_POS_UPDATE_Z);
+        setDesiredPosition(&targetHoldPos, posControl.actualState.yaw, NAV_POS_UPDATE_XY | NAV_POS_UPDATE_HEADING);
```
Comment (DzikuVx) [inline src/main/flight/navigation_rewrite.c:761]: This looks stange. Wouldn't it be enough to just assume posControl.actualState.pos as current XZY ?

**Comment by iforce2d** [comment]
Yes, I will test when weather allows.

Code looks like it makes sense, although I get the feeling that in the long term, it might be better to have the GCS as a proper mode in the state machine. Right now for my follow-me project I am doing things like looking at the current offset (lat,lon.alt) from the follow-me target when entering follow-me mode, and using that as the target offset for follow-me position. For example, if I position the quad 20m away from me toward the east, and 10m above, and I enter follow-me mode, the quad should remain 20m to the east and 10m above me as I move around. This requires extra house-keeping to be done at the point of changing modes, to record the target offsets.

Right now I am doing everything externally in my arduino program, which communicates via MSP to the flight controller. To detect when modes are changed, I read the rc inputs info a few times per second and check the values of the relevant stick channels - when I detect a change I know that follow-me mode has been entered or exited.

So there is no big deal and I can do what I need to for now, but if we wanted to have iNav include this kind of functionality, adding the GCS mode to the state machine might be easier in the long run.

File: src/main/flight/navigation_rewrite.c (hunk 1 above)
Comment (digitalentity) [inline src/main/flight/navigation_rewrite.c:761]: We might be moving that's why I recalculate `targetHoldPos` here.

**Comment by digitalentity** [comment]
@iforce2d I'm thinking about implementing GCS as a part of state machine for 1.3 and make it optional (selectable at compile time).

**Comment by iforce2d** [comment]
Cool, that sounds great.

While we're here, I found one other small change necessary to get the follow-me running correctly. In navigation_rewrite.c, function setWaypoint where WP 255 is handled, the yaw should be using centidegrees, not decidegrees.

Diff:

```
@@ -1820,7 +1911,7 @@ void setWaypoint(uint8_t wpNumber, navWaypoint_t * wpData)
             waypointUpdateFlags |= NAV_POS_UPDATE_HEADING;
         }

-        setDesiredPosition(&wpPos.pos, DEGREES_TO_DECIDEGREES(wpData->p1), waypointUpdateFlags);
+        setDesiredPosition(&wpPos.pos, DEGREES_TO_CENTIDEGREES(wpData->p1), waypointUpdateFlags);
     }
     // WP #1 - #15 - common waypoints - pre-programmed mission
     else if ((wpNumber >= 1) && (wpNumber <= NAV_MAX_WAYPOINTS) && !ARMING_FLAG(ARMED)) {
```

Sorry for not doing a proper pull request but git makes my head hurt. It usually takes me a demoralizing hour or more to figure out how to do what I want every time I use it. For this little thing so far I have not been able to muster the willpower.... eventually I will make a pull request if there is enough other stuff, but for now I'm avoiding the pain.

**Comment by digitalentity** [comment]
@iforce2d 
its already `setDesiredPosition(&wpPos.pos, DEGREES_TO_CENTIDEGREES(wpData->p1), waypointUpdateFlags);` in latest master.

**Comment by iforce2d** [comment]
aaaaarrggghhh..... I thought I was testing this today but now I see it is not part of the master branch. If the weather holds I will try again tomorrow.

**Comment by digitalentity** [comment]
@iforce2d I'll merge this right away. Please test master. If you find bugs we'll fix them on the go. Want to do an RC1 today :smile:

**Comment by iforce2d** [comment]
Thankyou :) Tested today, works well.

**Pull request merged**